producer/src/test/kotlin/streams/integrations/KafkaEventRouterLogCompactionTSE.kt

```diff
@@ -0,0 +1,627 @@
+import org.neo4j.internal.helpers.collection.Iterators
+import streams.extensions.execute
+import streams.events.*
+import streams.utils.JSONUtils
+import streams.setConfig
+import streams.start
+import java.time.Duration
+import java.util.*
+import kotlin.test.*
+
+class KafkaEventRouterLogCompactionTSE : KafkaEventRouterBaseTSE() {
+
+    private val bootstrapServerMap = mapOf("bootstrap.servers" to KafkaEventRouterSuiteIT.kafka.bootstrapServers)
+
+    private fun createCompactTopic(topic: String) = AdminClient.create(bootstrapServerMap).use {
+        val topics = listOf(compactTopic(topic))
+        it.createTopics(topics).all().get()
+    }
+
+    private fun compactTopic(topic: String) =
+            NewTopic(topic, 1, 1).configs(mapOf(
+                    "cleanup.policy" to "compact",
+                    "segment.ms" to "10",
+                    "retention.ms" to "1",
+                    "min.cleanable.dirty.ratio" to "0.01",
+            ))
+
+    private fun createProducerRecordKeyForDeleteStrategy(meta: Meta) = "${meta.txId + meta.txEventId}-${meta.txEventId}"
+
+    private fun initDbWithLogStrategy(strategy: String, otherConfigs: Map<String, String>? = null, constraints: List<String>? = null) {
+
+        db.setConfig("streams.source.schema.polling.interval", "0")
+                .setConfig("kafka.streams.log.compaction.strategy", strategy)
+
+        otherConfigs?.forEach { (k, v) -> db.setConfig(k, v) }
+        db.start()
+        constraints?.forEach { db.execute(it) }
+    }
+
+    private fun createManyPersons() = db.execute("UNWIND range(1, 9999) AS id CREATE (:Person {name:id})")
+
+    @Test
+    fun `compact message with streams publish`() {
+        val topic = UUID.randomUUID().toString()
+        initDbWithLogStrategy(TopicConfig.CLEANUP_POLICY_COMPACT)
+        createCompactTopic(topic)
+
+        KafkaEventRouterSuiteIT.registerPublishProcedure(db)
+        kafkaConsumer.subscribe(listOf(topic))
+
+        val keyRecord = "test"
+        // we sent 5 messages, 3 of them with the same key, so we expect that
+        // with the log compaction activated we expect to have just one message with the key equal to 'test'
+        db.execute("CALL streams.publish('$topic', 'Compaction 0', {key: 'Baz'})")
+        db.execute("CALL streams.publish('$topic', 'Compaction 1', {key: '$keyRecord'})")
+        db.execute("CALL streams.publish('$topic', 'Compaction 2', {key: '$keyRecord'})")
+        db.execute("CALL streams.publish('$topic', 'Compaction 3', {key: 'Foo'})")
+        db.execute("CALL streams.publish('$topic', 'Compaction 4', {key: '$keyRecord'})")
+
+        // to activate the log compaction process we publish dummy messages
+        db.execute("UNWIND range(1,9999) as id CALL streams.publish('$topic', 'id', {key: 'id'}) RETURN null") {
+            assertEquals(9999, Iterators.count(it))
+        }
+        // wait for messages population in topic
+        Thread.sleep(15000)
```

Comment:
your choice on how best to do this but my experience is whenever you have Thread.sleep in a test, the test ends up being flaky the one time kafka is too slow, or whatever.  Consider polling longer, until you get expected number of records.    This code pattern (Thread.sleep followed by poll) is repeated many times in these tests, you can refactor into a func to do it once in a more resilient way

Reply:
@vga91 please switch to `assertEventually`